awx/main/scheduler/task_manager.py

```diff
@@ -271,170 +270,107 @@ class DependencyManager(TaskBase):
-
-    def create_inventory_update(self, task, inventory_source_task):
-        inventory_task = InventorySource.objects.get(id=inventory_source_task.id).create_inventory_update(_eager_fields=dict(launch_type='dependency'))
-
-        inventory_task.created = task.created - timedelta(seconds=2)
-        inventory_task.status = 'pending'
-        inventory_task.save()
-        logger.debug('Spawned {} as dependency of {}'.format(inventory_task.log_format, task.log_format))
-
-        return inventory_task
-
-    def add_dependencies(self, task, dependencies):
-        with disable_activity_stream():
-            task.dependent_jobs.add(*dependencies)
-
-    def get_inventory_source_tasks(self):
+    def cache_projects_and_sources(self):
+        project_ids = set()
         inventory_ids = set()
         for task in self.all_tasks:
             if isinstance(task, Job):
-                inventory_ids.add(task.inventory_id)
-        self.all_inventory_sources = [invsrc for invsrc in InventorySource.objects.filter(inventory_id__in=inventory_ids, update_on_launch=True)]
-
-    def get_latest_inventory_update(self, inventory_source):
-        latest_inventory_update = InventoryUpdate.objects.filter(inventory_source=inventory_source).order_by("-created")
-        if not latest_inventory_update.exists():
-            return None
-        return latest_inventory_update.first()
-
-    def should_update_inventory_source(self, job, latest_inventory_update):
-        now = tz_now()
-
-        if latest_inventory_update is None:
-            return True
+                if task.project_id:
+                    project_ids.add(task.project_id)
+                if task.inventory_id:
+                    inventory_ids.add(task.inventory_id)
+            elif isinstance(task, InventoryUpdate):
+                if task.inventory_source and task.inventory_source.source_project_id:
+                    project_ids.add(task.inventory_source.source_project_id)
+
+        self.all_projects = {}
+        for proj in Project.objects.filter(id=project_ids, scm_update_on_launch=True):
+            self.all_projects[proj.id] = proj
+
+        self.all_inventory_sources = {}
+        for invsrc in InventorySource.objects.filter(inventory_id__in=inventory_ids, update_on_launch=True):
+            self.all_inventory_sources.setdefault(invsrc.inventory_id, [])
+            self.all_inventory_sources[invsrc.inventory_id].append(invsrc)
+
+    @staticmethod
+    def should_update_again(update, cache_timeout):
         '''
-        If there's already a inventory update utilizing this job that's about to run
-        then we don't need to create one
+        If it has never updated, we need to update
+        If there is already an update in progress then we do not need to a new create one
+        If the last update failed, we always need to try and update again
+        If current time is more than cache_timeout after last update, then we need a new one
         '''
```

Comment:
You can summarize the diff around here by saying that we combined 2 methods into 1. That is:

`should_update_inventory_source` & `should_update_related_project` --> `should_update_again`

the method signature changed slightly from `(self, job, latest_project_update/latest_inventory_update)` to `(update, cache_timeout)`. Maybe I should call it `latest_update` instead of `update` for clarity, and the fact that it's the same as the last parameter of the old method. This can be either a project update or a inventory update. And now that I think about it, it might be better to infer `cache_timeout` inside of this method, getting us down to 1 argument, because that is truly the only thing we need.